[REEF-1049]  Refactor Evaluator to properly use Tang Configuration

jwang98052 opened the pull request:
*Make the following constructor injectable:
  ContextManager
  HeartBeatManager
  EvaluatorRuntime
  EvaluatorSettings

*Consolidate injector for Clock, clrDriver Configuration and Evaluator
*modofied EvaluatorConfguration to properly use Tang configuration. It will be deprecated.
*Evaluator re-write and clean up

JIRA: [REEF-1049](https://issues.apache.org/jira/browse/REEF-1049)

This closes #

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Files/REEFFileNames.cs:
```diff
@@ -19,6 +19,7 @@
 
 using System.Diagnostics.CodeAnalysis;
 using System.IO;
+using System.Runtime.CompilerServices;
```
What for?

jwang98052 replied:
Removed

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Files/REEFFileNames.cs:
```diff
@@ -49,6 +50,7 @@ public sealed class REEFFileNames
         private const string DRIVER_CONFIGURATION_NAME = "driver.conf";
         private const string EVALUATOR_CONFIGURATION_NAME = "evaluator.conf";
         private const string CLR_DRIVER_CONFIGURATION_NAME = "clrdriver.conf";
+        private const string CLR_BRIDGE_CONFIGURSTION_NAME = "clrBridge.config";
```
Typo.

jwang98052 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Files/REEFFileNames.cs:
```diff
@@ -151,6 +153,11 @@ public string GetEvaluatorConfigurationPath()
             return EVALUATOR_CONFIGURATION_PATH;
         }
 
+        public string GetClrBridgeConfigurationName()
```
Please add documentation.

jwang98052 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextManager.cs:
```diff
@@ -33,18 +34,24 @@ namespace Org.Apache.REEF.Common.Runtime.Evaluator.Context
     public sealed class ContextManager : IDisposable
     {
-        private static readonly Logger LOGGER = Logger.GetLogger(typeof(ContextManager));
+        private static readonly Logger Logger = Logger.GetLogger(typeof(ContextManager));
```
This change seems unrelated to the goal at hand. Is this part of new style guides?

jwang98052 replied:
It did follow the styles. Since I touched this file, so made the change.

markusweimer replied:
This blows up this particular change a lot. Hence, it might be a good idea to move the rename of *all* `LOGGER` to `Logger` into a separate JIRA? That way, we can separate pure style changes from functionality changes in our commits.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/EvaluatorSettings.cs:
```diff
@@ -22,15 +22,26 @@
 
 namespace Org.Apache.REEF.Common.Runtime.Evaluator
 {
     // TODO: merge with EvaluatorConfigurations class
```
File a JIRA for this and make it a proper TODO while we are at it.

jwang98052 replied:
In fact, this PR resolved this issue. I have merged those two. EvaluatorConfigurations  is going to be obsolete. In my 3rd commit I made an hour ago, I have removed this TODO.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/EvaluatorSettings.cs:
```diff
@@ -39,7 +50,7 @@ public class EvaluatorSettings
         private readonly int _maxHeartbeatRetries;
 
-        private readonly ContextConfiguration _rootContextConfig;
+        ////private readonly ContextConfiguration _rootContextConfig;
```
Remove commented out code.

jwang98052 replied:
Removed in my 3rd commit pushed an hour ago. Please review on the latest.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/EvaluatorSettings.cs:
```diff
@@ -51,44 +62,47 @@ public class EvaluatorSettings
+            IRemoteManagerFactory remoteManagerFactory,
+            IInjector injector)
+
+        ////public EvaluatorSettings(
```
Remove commented out code.

jwang98052 replied:
Removed.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/EvaluatorSettings.cs:
```diff
@@ -51,44 +62,47 @@ public class EvaluatorSettings
             _clock = clock;
-            _remoteManager = remoteManager;
-            _injector = injecor;
+            _remoteManager = remoteManagerFactory.GetInstance(new REEFMessageCodec()); 
```
Shouldn't `REEFMessageCodec` be injectable as well?

jwang98052 replied:
Yes, it can be. Changed.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/EvaluatorSettings.cs:
```diff
@@ -137,13 +151,12 @@ public int MaxHeartbeatFailures
-                return _rootContextConfig;
-            }
-        }
+        public ContextConfiguration RootContextConfig { get; private set; }
```
Please document the attributes.

Also, do they need private setters? They are truly immutable, right? In which case it would be better to back them with a `readonly` field.

jwang98052 replied:
Done

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/EvaluatorSettings.cs:
```diff
@@ -174,12 +187,46 @@ public IRemoteManager<REEFMessage> RemoteManager
             }
         }
+
+        private Optional<TaskConfiguration> GetTaskConfiguration()
```
Why not a attribute for this?

jwang98052 replied:
You mean a property? It is really a method. I renamed it to avoid using Get. 
I was thinking to put RootServiceConfiguration as a named parameter in the constructor so that we don't need to GetNamedInstance here. But it is optional. Putting it in the constructor would cause injection exception if no value is provided in the configuration. In this method, I can use catch to skip the injection exception if the value is not given.

markusweimer replied:
Yes, I meant property.

Can't we have two constructors: One which takes the parameter and the other one which doesn't? Tang should call the most specific one possible, given the configuration.

jwang98052 replied:
Usually either service Configuration is given or task configuration is given when evaluate started. As both are string, overloading constructor cannot distinguish those two cases.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/EvaluatorSettings.cs:
```diff
@@ -174,12 +187,46 @@ public IRemoteManager<REEFMessage> RemoteManager
+                    new TaskConfiguration(taskConfigString));
+        }
+
+        private Optional<ServiceConfiguration> GetRootServiceConfiguration()
```
Why not an attribute for this?

jwang98052 replied:
Same as above

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/HeartBeatManager.cs:
```diff
@@ -84,32 +87,20 @@ public HeartBeatManager(EvaluatorSettings settings, IRemoteIdentifier remoteId)
 
-        [SuppressMessage("StyleCop.CSharp.NamingRules", "SA1307:AccessibleFieldsMustBeginWithUpperCaseLetter", Justification = "Intended to be private, exposed now before using future injection")]
-        public EvaluatorRuntime _evaluatorRuntime { get; set; }
+        public EvaluatorRuntime Evaluatorruntime { get; set; }
```
Shouldn't this be `EvaluatorRuntime`? Also, please document and consider make this `internal`.

jwang98052 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/HeartBeatManager.cs:
```diff
@@ -84,32 +87,20 @@ public HeartBeatManager(EvaluatorSettings settings, IRemoteIdentifier remoteId)
 
         [SuppressMessage("StyleCop.CSharp.NamingRules", "SA1307:AccessibleFieldsMustBeginWithUpperCaseLetter", Justification = "Intended to be private, exposed now before using future injection")]
-        public ContextManager _contextManager { get; set; }
+        public ContextManager Contextmanager { get; set; }
```
Same as above, `ContextManager`, document and `internal`.

jwang98052 replied:
Documented. 
It is used in Evalutor project so cannot be internal. We have REEF-1050 to set up friend relationship. I can modify it after merge with that PR.

jwang98052 commented:
@markusweimer I made 3rd commit at the same time you started review. Please refer to the latest.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/HeartBeatManager.cs:
```diff
@@ -84,32 +87,20 @@ public HeartBeatManager(EvaluatorSettings settings, IRemoteIdentifier remoteId)
-        public EvaluatorRuntime _evaluatorRuntime { get; set; }
+        public EvaluatorRuntime Evaluatorruntime { get; set; }
 
         [SuppressMessage("StyleCop.CSharp.NamingRules", "SA1307:AccessibleFieldsMustBeginWithUpperCaseLetter", Justification = "Intended to be private, exposed now before using future injection")]
```
This can be removed, right?

jwang98052 replied:
yes

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/HeartBeatManager.cs:
```diff
@@ -84,32 +87,20 @@ public HeartBeatManager(EvaluatorSettings settings, IRemoteIdentifier remoteId)
-                _evaluatorSettings = value;
-            }
-        }
+        public EvaluatorSettings EvaluatorSettings { get; private set; }
```
Consider making it `internal` as well as backed by a `readonly` field.

jwang98052 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/HeartBeatManager.cs:
```diff
@@ -84,32 +87,20 @@ public HeartBeatManager(EvaluatorSettings settings, IRemoteIdentifier remoteId)
             lock (_queuedHeartbeats)
             {
-                if (_evaluatorSettings.OperationState == EvaluatorOperationState.RECOVERY)
+                if (EvaluatorSettings.OperationState == EvaluatorOperationState.RECOVERY)
```
Is this referencing a `static` variable in `EvaluatorSettings`? That would be dangerous.

jwang98052 replied:
No, it is to reference to a property. I have changed to use _evaluatorSettings read only variable to avoid confusion.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Parameters/EvaluatorHeartbeatPeriodInMs.cs:
```diff
@@ -0,0 +1,28 @@
+namespace Org.Apache.REEF.Common.Runtime.Evaluator.Parameters
+{
+      [NamedParameter(Documentation = "Evaluator Heartbeat Period in ms", ShortName = "EvaluatorHeartbeatPeriodInMs", DefaultValue = "4000")]
+    public class EvaluatorHeartbeatPeriodInMs : Name<int>
```
maybe make this `internal`?

jwang98052 replied:
I can change it to internal now. One day if client wants to set it, we might have to either make it public or create some configuration module for it.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Parameters/HeartbeatMaxRetry.cs:
```diff
@@ -0,0 +1,28 @@
+namespace Org.Apache.REEF.Common.Runtime.Evaluator.Parameters
+{
+    [NamedParameter(Documentation = "Heartbeat Max Retry", ShortName = "HeartbeatMaxRetry", DefaultValue = "3")]
+    public class HeartbeatMaxRetry : Name<int>
```
`internal sealed`?

jwang98052 replied:
same comments as above

markusweimer commented on lang/cs/Org.Apache.REEF.Evaluator/Evaluator.cs:
```diff
@@ -47,80 +42,36 @@ namespace Org.Apache.REEF.Evaluator
+        private static Logger _logger = Logger.GetLogger(typeof(Evaluator));
+        private static IInjector _injector;
 
         public static void Main(string[] args)
```
Please document the command line parameters.

jwang98052 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.Evaluator/Evaluator.cs:
```diff
@@ -249,7 +202,8 @@ private static RuntimeClock InstantiateClock()
             IConfiguration clockConfiguration = new ConfigurationModuleBuilder().Build().Build();
```
Shouldn't the `Clock` be injected using the configuration provided for the Evaluator?

markusweimer commented:
I've done a surface level pass on this. Some global comments:

  * Please remove all commented out code added in this PR.
  * Why does this rename `LOGGER` to `Logger`? This seems purely stylistic and should probably go to a separate PR.
  * This makes incompatible changes to `public` members in `public` classes. I'm OK with that, as they should never have been `public` in the first place. Not sure how to deal with that. Maybe we need to declare 0.14 to be an incompatible release?

markusweimer commented:
@jwang98052 OK, I will do another pass.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/EvaluatorSettings.cs:
```diff
@@ -22,15 +22,25 @@
 
 namespace Org.Apache.REEF.Common.Runtime.Evaluator
 {
-    // TODO: merge with EvaluatorConfigurations class
```
This is still a TODO, right? Please file a JIRA and note it here.

jwang98052 replied:
I have merged EvaluatorConfigurations to EvaluatorSettigns in the PR and marked EvaluatorConfigurations as Obsolete. Created REEF-1068 to follow up it.

markusweimer commented on lang/cs/Org.Apache.REEF.Evaluator/Evaluator.cs:
```diff
@@ -47,84 +43,43 @@ namespace Org.Apache.REEF.Evaluator
+            clock.Run();
+        }
 
         public static void Main(string[] args)
```
Please document the command line parameters.

jwang98052 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.Evaluator/Evaluator.cs:
```diff
@@ -47,84 +43,43 @@ namespace Org.Apache.REEF.Evaluator
+
+        [Inject]
+        private Evaluator(
+            RuntimeClock clock,
```
Why `RuntimeClock` and not `IClock`?

jwang98052 replied:
The original code was to inject RuntimeClock directly. There is no default for IClock. I could set default of bind RuntimeClock  to IClock but all the methods used in those classes for Clock are all defined in RuntimeClock rather than in IClock, example is Run(). we might want to visit it separately.

markusweimer replied:
+1
Can you file a JIRA for that?

jwang98052 replied:
REEF-1069 is created.

markusweimer commented on lang/cs/Org.Apache.REEF.Evaluator/Evaluator.cs:
```diff
@@ -156,74 +111,22 @@ private static void AttachDebugger()
 
-                logger.Log(Level.Info, "Evaluator in debug mode, waiting for debugger to be attached...");
+                _logger.Log(Level.Info, "Evaluator in debug mode, waiting for debugger to be attached...");
                 Thread.Sleep(2000);
```
What is this sleep for?

jwang98052 replied:
This is original code. I am not sure. I guess it is for loading symbols for debug

markusweimer replied:
We should investigate this in a separate JIRA, then.

jwang98052 replied:
REEF-1070 opened.

markusweimer commented on lang/cs/Org.Apache.REEF.Evaluator/Evaluator.cs:
```diff
@@ -47,84 +43,43 @@ namespace Org.Apache.REEF.Evaluator
+            heartBeatManager.Evaluatorruntime = evaluatorRuntime;
+            heartBeatManager.Contextmanager = contextManager;
+            SetRuntimeHandlers(evaluatorRuntime, clock);
+            clock.Run();
```
Is this safe in C#? I'd expect a method `Run()` or such in `Evaluator` that is called in `Main`.

jwang98052 replied:
That is a good point. I also thought about it even there is only one line in the method. I will make a change.

markusweimer commented:
@jwang98052 Done with my 2nd pass.

jwang98052 commented:
Have addressed review comments and pushed in the changes. @markusweimer thanks for the review!  Please review again.

jwang98052 commented:
@markusweimer Please review it again and let me know when it is done. I guess I need to rebase and resolve conflict before the merge.

markusweimer commented:
I'll do another shallow pass. I can't test this as I currently don't have access to a Visual Studio 2013 box. Someone else will have to do the testing.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextManager.cs:
```diff
@@ -33,18 +34,24 @@ namespace Org.Apache.REEF.Common.Runtime.Evaluator.Context
 {
     public sealed class ContextManager : IDisposable
```
With [REEF-1050], this can be `internal`, right?

jwang98052 replied:
Yes, REEF-10560 has made it internal.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/EvaluatorRuntime.cs:
```diff
@@ -62,8 +62,8 @@ public sealed class EvaluatorRuntime : IObserver<RuntimeStart>, IObserver<Runtim
 
-                // start the hearbeat
-                _clock.ScheduleAlarm(0, heartBeatManager);
+                // start the hear beat
```
typo, `heart beat`

jwang98052 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/HeartBeatManager.cs:
```diff
@@ -37,9 +39,9 @@
 namespace Org.Apache.REEF.Common.Runtime.Evaluator
 {
-    public class HeartBeatManager : IObserver<Alarm>
+    public sealed class HeartBeatManager : IObserver<Alarm>
```
`internal`?

jwang98052 replied:
Already changed in REEF-1050. Will need to merge the change.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/HeartBeatManager.cs:
```diff
@@ -61,46 +61,50 @@ public class HeartBeatManager : IObserver<Alarm>
+        /// Return EvaluatorRuntime referenced from HeartBeatManager
+        /// Notes: As multiple tangling of the references, it cannot be injected in the constructor even using FutureInjection at this time
+        /// </summary>
+        public EvaluatorRuntime EvaluatorRuntime { get; set; }
```
Why is this mutable?

jwang98052 replied:
It is set in Evaluator after Evaluator gets EvaluatorRuntime. 

We could use inject future. However, we have triangle reference among EvaluatorRuntime, HeartBeatManager and ContextManager. I tried future injection for different combinations, it doesn't work.

jwang98052 replied:
Ok, I made future injection work. So the property is now mutable.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/HeartBeatManager.cs:
```diff
@@ -61,46 +61,50 @@ public class HeartBeatManager : IObserver<Alarm>
+        /// Return ContextManager referenced from HeartBeatManager
+        /// Notes: As multiple tangling of the references, it cannot be injected in the constructor even using FutureInjection at this time
+        /// </summary>
+        public ContextManager ContextManager { get; set; }
```
Why is this mutable?

jwang98052 replied:
same as above

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Parameters/EvaluatorHeartbeatPeriodInMs.cs:
```diff
@@ -0,0 +1,28 @@
+namespace Org.Apache.REEF.Common.Runtime.Evaluator.Parameters
+{
+    [NamedParameter(Documentation = "Evaluator Heartbeat Period in ms", ShortName = "EvaluatorHeartbeatPeriodInMs", DefaultValue = "4000")]
+    internal class EvaluatorHeartbeatPeriodInMs : Name<int>
```
`sealed`

jwang98052 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Parameters/HeartbeatMaxRetry.cs:
```diff
@@ -0,0 +1,28 @@
+namespace Org.Apache.REEF.Common.Runtime.Evaluator.Parameters
+{
+    [NamedParameter(Documentation = "Heartbeat Max Retry", ShortName = "HeartbeatMaxRetry", DefaultValue = "3")]
+    internal class HeartbeatMaxRetry : Name<int>
```
`sealed`

jwang98052 replied:
done

markusweimer commented:
I've done another surface-level pass. Overall, I have a hard time with this change. It would help a lot to first describe the desired end goal on the JIRA ([REEF-289](https://issues.apache.org/jira/browse/REEF-289)) and then to work towards that in a set of smaller changes.

That isn't to say this change doesn't make things better. It does. But it is unclear how it relates to the overall plan for the Evaluator. If someone else can test this, I think it can be merged.

jwang98052 commented:
I have update JIRA REEF-1049 with the following:
As a sub tack of REEF-289, this PR is mainly for using Tang injection properly for the following classes:
Evaluator
HeartBeatManager
Contextmanager
EvaluatorRuntime
EvaluatorSettings

Evaluator reference to EvaluatorRuntime
EvaluatorRuntime has ContextManager and HeartBeatManager

EvaluatorConfigurations is consolidated into EvaluatorSettings which contains all the deserailzed configurations data from evaluator config file, and ContextConfiguration, Optional<TaskConfiguration> and Optional<ServiceConfiguration> and Evaluator injector. 

The evaluator config file is passed into Evaluator exe as a command line parameter, it is deserialized in main program so that to get injector, this injector is used to inject Evaluator, then all nested object including EvaluatorSetting are injected. 

The logic change is out of scope of this Jira. The reference relationship among EvaluatorRuntime, Contextmanager and HeartBeatManager are also not changed in this Jira.

jwang98052 commented:
I have addressed review comments
Updated JIRA for the issues resolved in the PR
reverted Logger rename as it is not part of the PR and make tehe real change not clear. 
@markusweimer can you look at it again? I can do a rebase and merge from master if that makes review easier.

markusweimer commented:
Can you please resolve conflicts and rebase? I will test and merge then. After this PR, we should converge the design discussions on the JIRA.

jwang98052 commented:
@markusweimer rebased and pushed to the branch.

markusweimer commented:
I'll test and merge.
